### ISSUE TITLE
job-manager: cache inactive jobs for dependency resolution, etc

### ISSUE THREAD

**Issue by garlick** [flux-framework/flux-core#3702]
Problem: the simple dependency support added in #3696 add support for running after another job by ID.  If the after job is already inactive, the job manager no longer knows if the jobid was valid or the job result, so an exception is raised on the dependent job.

For example, it would be unreliable to submit a chain of fast-running, serially-dependent jobs if there is a chance of a job completing before its direct successor can be submitted.

Caching inactive jobs for some configurable period in the job manager could make this more reliable.

**Comment by grondo** [comment]
If this were only useful for the `dependency-after` plugin, then the plugin itself could implement this cache. However, it is presumed that other plugins may find it useful for their implementation to be able to lookup a possibly reduced set of information for an inactive job (e.g. userid, result), which is why it may be beneficial for the job-manager to implement this support.

**Comment by garlick** [comment]
Unable to leave well enough alone, I quickly hacked something together that puts jobs in an inactive hash instead of deleting them when they complete.  The code changes are really trivial (see below).  Inactive jobs are even reloaded on job manager restart without any changes to the replay code.

Note that the per job memory footprint should be fairly low here compared to job-list.  Here the jobspec is redacted of the environment, and only the terminating eventlog entry from the main eventlog is kept.

```diff
diff --git a/src/modules/job-manager/event.c b/src/modules/job-manager/event.c
index 2a3313444..a26e34352 100644
--- a/src/modules/job-manager/event.c
+++ b/src/modules/job-manager/event.c
@@ -371,6 +371,11 @@ int event_job_action (struct event *event, struct job *job)
         case FLUX_JOB_STATE_INACTIVE:
             if ((job->flags & FLUX_JOB_WAITABLE))
                 wait_notify_inactive (ctx->wait, job);
+            if (zhashx_insert (ctx->inactive_jobs, &job->id, job) < 0) {
+                flux_log_error (event->ctx->h,
+                                "%ju: error preserving inactive job",
+                                (uintmax_t) job->id);
+            }
             zhashx_delete (ctx->active_jobs, &job->id);
             drain_check (ctx->drain);
             break;
diff --git a/src/modules/job-manager/job-manager.c b/src/modules/job-manager/job-manager.c
index e4d0c83d7..4d24a0f29 100644
--- a/src/modules/job-manager/job-manager.c
+++ b/src/modules/job-manager/job-manager.c
@@ -144,12 +144,15 @@ int mod_main (flux_t *h, int argc, char **argv)
     memset (&ctx, 0, sizeof (ctx));
     ctx.h = h;
 
-    if (!(ctx.active_jobs = job_hash_create ())) {
-        flux_log_error (h, "error creating active_jobs hash");
+    if (!(ctx.active_jobs = job_hash_create ())
+        || !(ctx.inactive_jobs = job_hash_create ())) {
+        flux_log_error (h, "error creating jobs hash");
         goto done;
     }
     zhashx_set_destructor (ctx.active_jobs, job_destructor);
     zhashx_set_duplicator (ctx.active_jobs, job_duplicator);
+    zhashx_set_destructor (ctx.inactive_jobs, job_destructor);
+    zhashx_set_duplicator (ctx.inactive_jobs, job_duplicator);
     if (!(ctx.event = event_ctx_create (&ctx))) {
         flux_log_error (h, "error creating event batcher");
         goto done;
@@ -225,6 +228,7 @@ done:
     event_ctx_destroy (ctx.event);
     jobtap_destroy (ctx.jobtap);
     zhashx_destroy (&ctx.active_jobs);
+    zhashx_destroy (&ctx.inactive_jobs);
     return rc;
 }
 
diff --git a/src/modules/job-manager/job-manager.h b/src/modules/job-manager/job-manager.h
index 0a457de7e..b20469522 100644
--- a/src/modules/job-manager/job-manager.h
+++ b/src/modules/job-manager/job-manager.h
@@ -17,6 +17,7 @@ struct job_manager {
     flux_t *h;
     flux_msg_handler_t **handlers;
     zhashx_t *active_jobs;
+    zhashx_t *inactive_jobs;
     int running_jobs; // count of jobs in RUN | CLEANUP state
     flux_jobid_t max_jobid; // largest jobid allocated thus far
     struct start *start;
```

Branch is here:

https://github.com/garlick/flux-core/tree/keep_inactives

(based on top of #3696)

One perhaps naive thought is to just go with this as the default and in the system instance, arrange to purge (and archive) jobs from both the KVS and the job manager in one go - that way they always remain in sync.

But probably some more careful design consideration is needed here, and possibly some testing to see if this affects job throughput over time.

**Comment by garlick** [comment]
Some throughput tests that don't really show any impact that I'm spotting, FWIW:

With inactive job capture:
```
number of jobs: 40000
submit time:    10.953s (3651.9 job/s)
script runtime: 1183.668s
job runtime:    217.505s
throughput:     183.9 job/s (script:  33.8 job/s)

number of jobs: 40000
submit time:    11.020s (3629.8 job/s)
script runtime: 1127.179s
job runtime:    209.838s
throughput:     190.6 job/s (script:  35.5 job/s)

number of jobs: 1024
submit time:    0.525 s (1951.1 job/s)
script runtime: 3.662 s
job runtime:    3.433 s
throughput:     298.3 job/s (script: 279.7 job/s)

number of jobs: 1024
submit time:    0.582 s (1759.2 job/s)
script runtime: 3.331 s
job runtime:    3.106 s
throughput:     329.7 job/s (script: 307.4 job/s)

number of jobs: 1024
submit time:    0.317 s (3233.1 job/s)
script runtime: 3.404 s
job runtime:    3.389 s
throughput:     302.1 job/s (script: 300.8 job/s

number of jobs: 4096
submit time:    1.169 s (3502.5 job/s)
script runtime: 14.622s
job runtime:    14.449s
throughput:     283.5 job/s (script: 280.1 job/s)

number of jobs: 4096
submit time:    1.331 s (3076.3 job/s)
script runtime: 17.523s
job runtime:    17.364s
throughput:     235.9 job/s (script: 233.7 job/s)

number of jobs: 4096
submit time:    1.323 s (3095.3 job/s)
script runtime: 18.130s
job runtime:    17.987s
throughput:     227.7 job/s (script: 225.9 job/s)
```
Without
```
number of jobs: 40000
submit time:    10.897s (3670.9 job/s)
script runtime: 1154.583s
job runtime:    212.699s
throughput:     188.1 job/s (script:  34.6 job/s)

number of jobs: 1024
submit time:    0.388 s (2636.5 job/s)
script runtime: 3.136 s
job runtime:    3.036 s
throughput:     337.3 job/s (script: 326.5 job/s)

number of jobs: 1024
submit time:    0.328 s (3126.0 job/s)
script runtime: 3.428 s
job runtime:    3.415 s
throughput:     299.9 job/s (script: 298.7 job/s)

number of jobs: 1024
submit time:    0.338 s (3031.1 job/s)
script runtime: 3.403 s
job runtime:    3.393 s
throughput:     301.8 job/s (script: 300.9 job/s)

number of jobs: 4096
submit time:    1.244 s (3293.9 job/s)
script runtime: 18.269s
job runtime:    18.075s
throughput:     226.6 job/s (script: 224.2 job/s)

number of jobs: 4096
submit time:    1.441 s (2841.6 job/s)
script runtime: 17.492s
job runtime:    17.360s
throughput:     236.0 job/s (script: 234.2 job/s)

submit time:    1.379 s (2969.6 job/s)
script runtime: 17.612s
job runtime:    17.428s
throughput:     235.0 job/s (script: 232.6 job/s)
```

**Comment by grondo** [comment]
This should have been closed by #4286
